src/nsrt_learning/strips_learning/clustering_learner.py

```diff
@@ -129,8 +142,168 @@ class ClusterAndSearchSTRIPSLearner(ClusteringSTRIPSLearner):
     """A clustering STRIPS learner that learns preconditions via search,
     following the LOFT algorithm: https://arxiv.org/abs/2103.00589."""
 
-    def _learn_pnad_preconditions(self, pnad: PartialNSRTAndDatastore) -> None:
-        raise NotImplementedError
+    def _learn_pnad_preconditions(
+            self, pnads: List[PartialNSRTAndDatastore]
+    ) -> List[PartialNSRTAndDatastore]:
+        new_pnads = []
+        for i, pnad in enumerate(pnads):
+            positive_data = pnad.datastore
+            # Construct negative data by merging the datastores of all
+            # other PNADs that have the same option.
+            negative_data = []
+            for j, other_pnad in enumerate(pnads):
+                if i == j:
+                    continue
+                if pnad.option_spec[0] != other_pnad.option_spec[0]:
+                    continue
+                negative_data.extend(other_pnad.datastore)
+            all_preconditions = self._run_search(pnad, positive_data,
+                                                 negative_data)
+            # The datastores of the new PNADs could need to be changed, so
+            # we initialize them as empty, then recompute them at the end.
+            for j, preconditions in enumerate(all_preconditions):
+                new_pnads.append(
+                    PartialNSRTAndDatastore(
+                        pnad.op.copy_with(name=f"{pnad.op.name}-{j}",
+                                          preconditions=preconditions), [],
+                        pnad.option_spec))
+        self._recompute_datastores_from_segments(new_pnads)
+        return new_pnads
+
+    def _run_search(self, pnad: PartialNSRTAndDatastore,
+                    positive_data: Datastore,
+                    negative_data: Datastore) -> Set[FrozenSet[LiftedAtom]]:
+        """Run outer-level search to find a set of precondition sets.
+
+        Each precondition set will produce one operator.
+        """
+        all_preconditions = set()
+        # We'll remove positives as they get covered.
+        remaining_positives = list(positive_data)
+        while remaining_positives:
+            new_preconditions = self._run_single_search(
+                pnad, remaining_positives, negative_data)
+            # Update the remaining positives.
+            new_remaining_positives = []
+            for seg, var_to_obj in remaining_positives:
+                ground_pre = {a.ground(var_to_obj) for a in new_preconditions}
+                if not ground_pre.issubset(seg.init_atoms):
+                    # If the preconditions ground with this substitution don't
+                    # hold in this segment's init_atoms, this segment has yet
+                    # to be covered, so we keep it in the positives.
+                    new_remaining_positives.append((seg, var_to_obj))
+                else:
+                    # Otherwise, we can move this segment to negative_data,
```

Comment:
did we do this before? does it matter? what's the justification?

Reply:
we did it before yeah, so that's my justification :P staying true to the loft implementation. i don't think it matters in our current envs